### ISSUE TITLE
stack overflow exception in tomcat with security manager

### ISSUE THREAD

**Issue by hdeadman** [mjiderhamn/classloader-leak-prevention#55]
When I run the latest 1.x release of this listener (I have listener that extends it and overrides logger methods) I get a stack overflow exception if I am running tomcat with a security manager policy (catalina.sh start -security).  

-security adds these options to java:
-Djava.security.manager -Djava.security.policy==/var/lib/tomcats/clustera/conf/catalina.policy

I will look into changing the policy file to let this listener do whatever it wants but is there any way the listener could avoid getting stuck in this loop?

```
Aug 29, 2016 9:42:56 AM org.apache.catalina.core.ContainerBase.[Catalina].[localhost].[/xyz##1.2.3] listenerStart
SEVERE: Exception sending context initialized event to listener instance of class org.xyz.web.memleak.ClassloaderMemoryLeakAvoidanceListener
java.lang.StackOverflowError
        at java.security.AccessControlContext.optimize(AccessControlContext.java:608)
        at java.security.AccessController.checkPermission(AccessController.java:883)
        at java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
        at java.lang.ClassLoader.checkClassLoaderPermission(ClassLoader.java:1528)
        at java.lang.ClassLoader.getParent(ClassLoader.java:1374)
        at se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor.isWebAppClassLoaderOrChild(ClassLoaderLeakPreventor.java:1803)
        at se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor$1.combine(ClassLoaderLeakPreventor.java:212)
        at java.security.AccessControlContext.optimize(AccessControlContext.java:608)
        at java.security.AccessController.checkPermission(AccessController.java:883)
        at java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
        at java.lang.ClassLoader.checkClassLoaderPermission(ClassLoader.java:1528)
        at java.lang.ClassLoader.getParent(ClassLoader.java:1374)
        at se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor.isWebAppClassLoaderOrChild(ClassLoaderLeakPreventor.java:1803)
        at se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor$1.combine(ClassLoaderLeakPreventor.java:212)
        at java.security.AccessControlContext.optimize(AccessControlContext.java:608)
        at java.security.AccessController.checkPermission(AccessController.java:883)
        at java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
        at java.lang.ClassLoader.checkClassLoaderPermission(ClassLoader.java:1528)
        at java.lang.ClassLoader.getParent(ClassLoader.java:1374)
        at se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor.isWebAppClassLoaderOrChild(ClassLoaderLeakPreventor.java:1803)
        at se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor$1.combine(ClassLoaderLeakPreventor.java:212)
        at java.security.AccessControlContext.optimize(AccessControlContext.java:608)
        at java.security.AccessController.checkPermission(AccessController.java:883)
        at java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
        at java.lang.ClassLoader.checkClassLoaderPermission(ClassLoader.java:1528)
        at java.lang.ClassLoader.getParent(ClassLoader.java:1374)
etc, etc, etc,...
```

**Comment by hdeadman** [comment]
If I run tomcat without the security manager but I do -Djava.security.debug="access", then I see the following logged twice:

access: AccessControlContext invoking the Combiner

If I run with a security manager and a policy that includes an entry that should let the webapps have AllPermission to anything, the stack overflow still happens and I see this logging from the -Djava.security.debug="access" as follows:

```
...
access: access allowed ("java.util.logging.LoggingPermission" "control")
access: access allowed ("java.util.logging.LoggingPermission" "control")
access: access allowed ("java.util.logging.LoggingPermission" "control")
access: access allowed ("java.util.logging.LoggingPermission" "control")
access: access allowed ("java.lang.RuntimePermission" "createClassLoader")
access: access allowed ("java.lang.reflect.ReflectPermission" "suppressAccessChecks")
access: access allowed ("java.io.FilePermission" "/var/lib/tomcats/clusterb/webapps/xyz##1.2.3/WEB-INF/classes/org/jboss" "read")
access: access allowed ("java.util.PropertyPermission" "java.vendor" "read")
access: access allowed ("java.io.FilePermission" "/var/lib/tomcats/clusterb/webapps/xyz##1.2.3/WEB-INF/classes/se/jiderhamn/classloader/leak/prevention/ClassLoaderLeakPreventor$2.class" "read")
access: access allowed ("java.lang.RuntimePermission" "getClassLoader")
access: access allowed ("java.lang.RuntimePermission" "setContextClassLoader")
access: access allowed ("java.io.FilePermission" "/var/lib/tomcats/clusterb/webapps/xyz##1.2.3/WEB-INF/classes/se/jiderhamn/classloader/leak/prevention/ClassLoaderLeakPreventor$3.class" "read")
access: access allowed ("java.security.SecurityPermission" "createAccessControlContext")
access: AccessControlContext invoking the Combiner
access: AccessControlContext invoking the Combiner
access: AccessControlContext invoking the Combiner
access: AccessControlContext invoking the Combiner
access: AccessControlContext invoking the Combiner
access: AccessControlContext invoking the Combiner
access: AccessControlContext invoking the Combiner
access: AccessControlContext invoking the Combiner
access: AccessControlContext invoking the Combiner
access: AccessControlContext invoking the Combiner
access: AccessControlContext invoking the Combiner
access: AccessControlContext invoking the Combiner
access: AccessControlContext invoking the Combiner
... This repeats for a long time, before stack overflow exception
```

**Comment by mjiderhamn** [comment]
Thanks for the report @hdeadman. Could you build and try the [`SecurityConstants.GET_CLASSLOADER_PERMISSION` branch](https://github.com/mjiderhamn/classloader-leak-prevention/tree/SecurityConstants.GET_CLASSLOADER_PERMISSION)?

The problem is the recursion in `isWebAppClassLoaderOrChild()`. On the branch above I've tried to work around that by calling the package protected `ClassLoader.isAncestor()` instead, but I'm not sure how well that plays with Jigsaw in JDK 9. If that doesn't work, we'd probably need to disable recursion, at least in nested `combine()` calls.

**Comment by hdeadman** [comment]
That fixes the error on deploy but I got the following error on undeploy: 

```
[ERROR] [2016.08.30.08:45:11] [  ] [org.apache.catalina.core.ContainerBase.[Catalina].[localhost].[/xyz##1.2.3]] Exception sending context destroyed event to listener instance of class org.xyz.web.memleak.ClassloaderMemoryLeakAvoidanceListener
java.lang.ClassCircularityError: java/lang/ClassLoaderUtils
        at se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor.isWebAppClassLoaderOrChild(ClassLoaderLeakPreventor.java:1801)
        at se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor.isLoadedByWebApplication(ClassLoaderLeakPreventor.java:1793)
        at se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor.isLoadedInWebApplication(ClassLoaderLeakPreventor.java:1788)
        at se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor$4.filter(ClassLoaderLeakPreventor.java:878)
        at javax.imageio.spi.FilterIterator.advance(ServiceRegistry.java:821)
        at javax.imageio.spi.FilterIterator.<init>(ServiceRegistry.java:815)
        at javax.imageio.spi.ServiceRegistry.getServiceProviders(ServiceRegistry.java:516)
        at se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor.deregisterIIOServiceProvider(ClassLoaderLeakPreventor.java:884)
        at se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor.contextDestroyed(ClassLoaderLeakPreventor.java:776)
        at org.xyz.web.memleak.ClassloaderMemoryLeakAvoidanceListener.contextDestroyed(ClassloaderMemoryLeakAvoidanceListener.java:23)
        at org.apache.catalina.core.StandardContext.listenerStop(StandardContext.java:4889)
        at org.apache.catalina.core.StandardContext.stopInternal(StandardContext.java:5526)
        at org.apache.catalina.util.LifecycleBase.stop(LifecycleBase.java:224)
        at org.apache.catalina.core.ContainerBase$StopChild.call(ContainerBase.java:1423)
        at org.apache.catalina.core.ContainerBase$StopChild.call(ContainerBase.java:1412)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

**Comment by mjiderhamn** [comment]
Never seen that one before, but it could be that there is some mixup when the `ClassLoaderUtils` is first loaded by the system classloader, and then needed by the web app classloader. Could you pull the latest changes and give it another try?

**Comment by hdeadman** [comment]
OK, I will try, but I see the unit tests failing with a SecurityException due to declaring a class in the java.lang.* package. I wonder why it is allowed in Tomcat?

**Comment by hdeadman** [comment]
With that change I am getting the same exception as the unit tests. 

```
SEVERE: Exception sending context initialized event to listener instance of class org.xyz.web.memleak.ClassloaderMemoryLeakAvoidanceListener
java.lang.SecurityException: Prohibited package name: java.lang
        at java.lang.ClassLoader.preDefineClass(ClassLoader.java:662)
        at java.lang.ClassLoader.defineClass(ClassLoader.java:761)
        at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
        at org.apache.catalina.loader.WebappClassLoaderBase.findClassInternal(WebappClassLoaderBase.java:2541)
        at org.apache.catalina.loader.WebappClassLoaderBase.findClass(WebappClassLoaderBase.java:858)
        at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1301)
        at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1166)
        at se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor.contextInitialized(ClassLoaderLeakPreventor.java:286)
        at org.xyz.web.memleak.ClassloaderMemoryLeakAvoidanceListener.contextInitialized(ClassloaderMemoryLeakAvoidanceListener.java:19)
        at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4842)
        at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5303)
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:147)
        at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:725)
        at org.apache.catalina.core.ContainerBase.access$000(ContainerBase.java:131)
        at org.apache.catalina.core.ContainerBase$PrivilegedAddChild.run(ContainerBase.java:153)
        at org.apache.catalina.core.ContainerBase$PrivilegedAddChild.run(ContainerBase.java:143)
        at java.security.AccessController.doPrivileged(Native Method)
        at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:699)
        at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:717)
        at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:940)
        at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:467)
        at org.apache.catalina.startup.HostConfig.check(HostConfig.java:1612)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.tomcat.util.modeler.BaseModelMBean.invoke(BaseModelMBean.java:300)
        at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.invoke(DefaultMBeanServerInterceptor.java:819)
        at com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:801)
        at org.apache.catalina.ha.deploy.FarmWarDeployer.check(FarmWarDeployer.java:587)
        at org.apache.catalina.ha.deploy.FarmWarDeployer.fileModified(FarmWarDeployer.java:467)
        at org.apache.catalina.ha.deploy.WarWatcher.check(WarWatcher.java:100)
        at org.apache.catalina.ha.deploy.FarmWarDeployer.backgroundProcess(FarmWarDeployer.java:571)
        at org.apache.catalina.ha.tcp.SimpleTcpCluster.backgroundProcess(SimpleTcpCluster.java:497)
        at org.apache.catalina.core.ContainerBase.backgroundProcess(ContainerBase.java:1117)
        at org.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.processChildren(ContainerBase.java:1376)
        at org.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.processChildren(ContainerBase.java:1380)
        at org.apache.catalina.core.ContainerBase$ContainerBackgroundProcessor.run(ContainerBase.java:1348)
        at java.lang.Thread.run(Thread.java:745)
```

**Comment by mjiderhamn** [comment]
I was afraid of that. New approach - could you give the [`NestedProtectionDomainCombinerException` branch](https://github.com/mjiderhamn/classloader-leak-prevention/tree/NestedProtectionDomainCombinerException) a try instead?

**Comment by hdeadman** [comment]
That appears to have fixed the issues I was seeing, no errors on deploy or undeploy. 

Thank you very much.

**Comment by mjiderhamn** [comment]
I have now released 1.15.4 and 2.0.1 with these changes.
